pkg/operators/klusterlet/controllers/ssarcontroller/klusterlet_ssar_controller.go

```diff
@@ -0,0 +1,429 @@
+package ssarcontroller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	authorizationv1 "k8s.io/api/authorization/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	coreinformer "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	corelister "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	operatorv1client "open-cluster-management.io/api/client/operator/clientset/versioned/typed/operator/v1"
+	operatorinformer "open-cluster-management.io/api/client/operator/informers/externalversions/operator/v1"
+	operatorlister "open-cluster-management.io/api/client/operator/listers/operator/v1"
+	"open-cluster-management.io/registration-operator/pkg/helpers"
+)
+
+type ssarController struct {
+	kubeClient       kubernetes.Interface
+	secretLister     corelister.SecretLister
+	klusterletClient operatorv1client.KlusterletInterface
+	klusterletLister operatorlister.KlusterletLister
+	sync.RWMutex
```

Comment:
i think you would want a lock for each klusterlet

Reply:
A little confused about this, it doesn't seem different klusterlet would have an effect on each other.

Reply:
when you have another klusterlet comes in, you will get hung when add, right?

Reply:
It won't hang because the lock only happens in the internal of `inSSARChecking`, `addSSARChecking`, `deleteSSARChecking`

When cluster1 is processing, the t `addSSARChecking`, `deleteSSARChecking` are not calling obviously. 

The `inSSARChecking` is called when a new reconcile begins, but **immediately unlocked** and returns a result (for same klusterlet return true and for another no-processing klusterlet return false) and there would be no hanging stage.

Reply:
Separate the locker into an independent struct and add a test function `TestLocker` to simulate the situation.